torch/_inductor/triton_heuristics.py

```diff
@@ -211,17 +216,26 @@ def precompile(self, warm_cache_only_with_cc=None):
                     # Note both A100 and H100 have 65536 32-bit registers per SM.
                     if nreg <= 65536 // device_prop.max_threads_per_multi_processor:
                         continue
-                    max_blocks_per_sm = device_prop.max_threads_per_multi_processor // (
-                        32 * triton_config.num_warps
-                    )
+
+                    nreg_per_warp = nreg * 32
+                    nreg_per_block = nreg_per_warp * triton_config.num_warps
+
+                    # Previously we set max_blocks_per_sm to 'max_threads_per_multi_processo / (32 * num_warps)'
+                    # The formula below is a tighter upper bound since we have the assumption that
+                    #   nreg > 65536 // device_prop.max_threads_per_multi_processor
+                    # due to the if condition above and:
+                    #   65536 / nreg_per_block
+                    #   = 65536 / (nreg * 32 * num_warps)
+                    #   < 65536 / ((65536 / max_threads_per_multi_processor) * 32 * num_warps)
+                    #   = max_threads_per_multi_processor / (32 * num_warps)
+                    # Using a tigher upper bound can reveal more optimization opportunities.
+                    max_blocks_per_sm = max(65536 // nreg_per_block, 1)
```

Comment:
> Here, if 65536 // nreg_per_block < 1 there is going to be register spillage, so we probably want to decrease the RBLOCK as well.

So I use `max(65536 // nreg_per_block, 1)` just to be safe. nreg_per_block actually is the register usage after considering register spills. The spilled registers are not counted in nreg_per_block . So if `65536 // nreg_per_block < 1` probably the kernel will not able to be run. 

Register spill is implicitly handled here actually. When register spills, each thread gonna uses 255 registers (the maximum # of register a thread can use). In that case, we have large chance to trigger the decreasing or RBLOCK.

Reply:
> max_blocks_per_sm = min(65536 // nreg_per_block, max_warps_per_sm // triton_conf.num_warps)

This is equivalent to the current code. My comment in the code shows why the following is true:
```
65536 // nreg_per_block < max_warps_per_sm // triton_conf.num_warps
```
here.

Scaling up RBLOCK can do harm since we increase register usage. Even if there is no register spills,  large register usage can still hurt occupancy and may result in under-performing configs.

Reply:
> When register spills, each thread gonna uses 255 registers (the maximum # of register a thread can use). In that case, we have large chance to trigger the decreasing of RBLOCK.

Why don't we handle it explicitly now that we are at it?